DeviceCode/Targets/Native/STM32F4/DeviceCode/Initialization/tinyhal.cpp

```diff
@@ -0,0 +1,752 @@
+                    break;
+
+                case ConfigurationSector::c_BootEntryKey: // looks like we already have a key set
+                    bDone = true;
+                    break;
+            }
+        }
+
+        if(!bDone) // we must be full, so rewrite sector
+        {
+            // reading whole block, not just the configurationsector
+            const BlockRegionInfo * pBlockRegionInfo = &deviceInfo->Regions[iRegion];
+
+            ::Watchdog_ResetCounter();
+
+            BYTE *data  = (BYTE*)  private_malloc(pBlockRegionInfo->BytesPerBlock);
+
+            if(data != NULL)
+            {
+                ConfigurationSector *pCfg;
+
+                if(deviceInfo->Attribute.SupportsXIP)
+                {
+                    memcpy(data, (void*)configSectAddress, pBlockRegionInfo->BytesPerBlock);
+                }
+                else
+                {
+                    pBlockDevice->Read(configSectAddress, pBlockRegionInfo->BytesPerBlock, data);
+                }
+
+                pCfg = (ConfigurationSector *)data;
+                memset( (void*)&pCfg->BooterFlagArray[0], 0xFF, sizeof(pCfg->BooterFlagArray) );
+
+                // updated the g_ConfigurationSector with the latest value
+                // as the g_ConfigurationSector must be at the RAM area, so it should be ok to write to it.
+                pCfg->BooterFlagArray[0]  = c_Key;
+
+                pBlockDevice->EraseBlock(configSectAddress);
+
+                ::Watchdog_ResetCounter();
+
+                // write back to sector, as we only change one bit from 0 to 1, no need to erase sector
+                bRet = (TRUE == pBlockDevice->Write( configSectAddress, pBlockRegionInfo->BytesPerBlock, data, FALSE ));
+
+                private_free(data);
+            }
+        }
+
+        CPU_Reset();
+    }
+}
+
+bool g_fDoNotUninitializeDebuggerPort = false;
+
+void HAL_Initialize()
+{    
+#if defined(PLATFORM_ARM_OS_PORT)
+    // Interrupts must be enabled to handle calls to OS
+    // (Network stack uses the CMSIS-RTX OS, which uses
+    // SVC calls, which will hard fault if the interrupts
+    // are disabled at the Svc instruction )
+    // SystemInit handles this for the startup from reset
+    // However, this is also called from the CLR when doing
+    // a soft reboot.
+    __enable_irq();
+#endif
+	
+    HAL_CONTINUATION::InitializeList();
+    HAL_COMPLETION  ::InitializeList();
+
+    HAL_Init_Custom_Heap();
+
+    Time_Initialize();
+    Events_Initialize();
+
+    CPU_GPIO_Initialize();
+    CPU_SPI_Initialize();
+
+#if !defined(PLATFORM_ARM_OS_PORT)
+    // this is the place where interrupts are enabled after boot for the first time after boot
+    ENABLE_INTERRUPTS();
+#endif
+
+    // have to initialize the blockstorage first, as the USB device needs to update the configure block
+
+    BlockStorageList::Initialize();
+
+    BlockStorage_AddDevices();
+
+    BlockStorageList::InitializeDevices();
+
+    //FS_Initialize();

```

Comment:
These shouldn't be commented out here in a general solution as they will never be called, even if the related features are added to the TinyCLR.proj. Ultimately we need a better way to leave these out when not needed so they don't force the inclusion of other dependencies and stubs, etc...